exonum-java-binding-core/src/test/java/com/exonum/binding/storage/indices/CheckedMapProofMultiMatcher.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.storage.indices;
+
+import com.exonum.binding.common.proofs.map.flat.CheckedMapProof;
+import com.exonum.binding.common.proofs.map.flat.MapEntry;
+import com.exonum.binding.common.proofs.map.flat.ProofStatus;
+import com.exonum.binding.common.serialization.StandardSerializers;
+import com.google.common.io.BaseEncoding;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+import org.hamcrest.core.IsEqual;
+
+class CheckedMapProofMultiMatcher extends TypeSafeMatcher<CheckedMapProof> {
+
+  private static final BaseEncoding HEX_ENCODING = BaseEncoding.base16().lowerCase();
+
+  private final List<MapTestEntry> entries;
+
+  private CheckedMapProofMultiMatcher(List<MapTestEntry> entries) {
+    this.entries = entries;
+  }
+
+  @Override
+  protected boolean matchesSafely(CheckedMapProof checkedMapProof) {
+    ProofStatus status = checkedMapProof.getStatus();
+    return status == ProofStatus.CORRECT
+        && entries.stream().allMatch(e -> checkEntry(checkedMapProof, e));
+  }
+
+  private boolean checkEntry(CheckedMapProof checkedMapProof, MapTestEntry entry) {
```

Comment:
I think it's high time we add a simple ByteString and use it in CheckedMapProof. If we had this, this class could look like that:

```java

  private CheckedMapProofMultiMatcher(Collection<MapTestEntry> entries) {
    Set<ByteString> expectedMissingKeys = /* filter entries by present/absent attribute 
                                             and take absent, converting them if appropriate */
    Set<RawMapEnty> expectedEntries = /* filter entries by present/absent attribute 
                                         and take present, converting them if appropriate */
    this.missingKeysMatcher = IsEqual.equalTo(expectedMissingKeys);
    this.entriesMatcher = IsEqual.equalTo(expectedEntries);
  }

  // That's a complete matchesSafely implementation!
  protected boolean matchesSafely(CheckedMapProof checkedMapProof) {
    ProofStatus status = checkedMapProof.getStatus();
    return status == ProofStatus.CORRECT
        && missingKeysMatcher.matches(checkedMapProof.getMissingKeys())
        && entriesMatcher.matches(checkedMapProof.getEntries());
  }
```

If you'd rather not make these changes at this moment — please document in an issue and add a TODO referencing it.

Reply:
I'll create a task 👍

Reply:
Please add the snippet above as one of the motivations.